### PR TITLE
JIT: Fold more always failing type checks

### LINKED ISSUE #97000: Array variance checks are not optimized away for types known to be not arrays
### Description

It appears that array variance checks against e.g. `List<T>` are not optimized away for patterns like `(obj)list is int[]`.

### Configuration

.NET 8.0.100

### Regression?

No

### Analysis

Given below pattern
```cs
static bool Foo(List<int> l)
{
    return (object)l is int[];
}
```
The produced asm is
```asm
G_M39343_IG01:  ;; offset=0x0000
       push     rax
G_M39343_IG02:  ;; offset=0x0001
       mov      rsi, rdi
       mov      rdi, 0x7FDDF17D91D0      ; int[]
       call     [CORINFO_HELP_ISINSTANCEOFARRAY]
       test     rax, rax
       setne    al
       movzx    rax, al
G_M39343_IG03:  ;; offset=0x001D
       add      rsp, 8
       ret      
```
instead of
```asm
       xor rax, rax
       ret
```

Should this pattern get optimized, this will allow to update e.g. `Enumerable.SequenceEquals` in such a way that type tests are optimized away and it directly forwards to `CollectionsMarshal.AsSpan(list).SequenceEquals(...)`.

### PR DESCRIPTION
**Pull request by jakobbotsch**
Folding `x isinst B` where `x` is statically of type `A` is currently done via a call to `compareTypesForCast`. If it returns `TypeCompareState::Must` then we know that `A` can be cast to `B`, which also means any class derived from `A` can be cast to `B`, so we can always fold that. However, if it returns `TypeCompareState::MustNot` we only learn something if we know `x` is exactly `A`; otherwise, if `x` is more derived than `A`, it could still potentially be cast to `B`.

In cases where the type hierarchies are disjoint, however, we can still know that some type checks will fail. To detect the case add a query in the opposite direction; that is, if `B` cannot be cast to `A`, then the type hierarchies are disjoint, so in that case we can fold it, regardless of having exact knowledge. This also requires some explicit checks for interface types, but perhaps we can move those to the EE side.

Fix #97000

```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static bool Test(List<int> x) => Foo(x);

[MethodImpl(MethodImplOptions.AggressiveInlining)]
static bool Foo<TSource>(IEnumerable<TSource> x) => x is TSource[];
```
Before:
```asm
G_M10367_IG01:  ;; offset=0x0000
       4883EC28             sub      rsp, 40
						;; size=4 bbWeight=1 PerfScore 0.25

G_M10367_IG02:  ;; offset=0x0004
       488BD1               mov      rdx, rcx
       48B9E88DCCB6FB7F0000 mov      rcx, 0x7FFBB6CC8DE8      ; int[]
       E8EAF6165E           call     CORINFO_HELP_ISINSTANCEOFARRAY
       4885C0               test     rax, rax
       0F95C0               setne    al
       0FB6C0               movzx    rax, al
						;; size=27 bbWeight=1 PerfScore 3.00

G_M10367_IG03:  ;; offset=0x001F
       4883C428             add      rsp, 40
       C3                   ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 36
```
After:
```asm
G_M10367_IG02:  ;; offset=0x0000
       33C0                 xor      eax, eax
						;; size=2 bbWeight=1 PerfScore 0.25

G_M10367_IG03:  ;; offset=0x0002
       C3                   ret      
						;; size=1 bbWeight=1 PerfScore 1.00
; Total bytes of code: 3
```

### CODE REVIEW

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -5303,70 +5303,83 @@ GenTree* Compiler::impOptimizeCastClassOrIsInst(GenTree* op1, CORINFO_RESOLVED_T
...
+
+    // We know that an object of type fromClass cannot be cast to toClass, but
+    // if it is more derived it could still succeed a cast.
+    if (!isExact)
+    {
+        // We can still fail some type casts if they are between two types that
+        // belong to disjoint type hierarchies.
+        // TODO: compareTypesForCast(interfaceType, otherType) returning
+        // MustNot seems less useful than just having it return May. If we
+        // change this on EE side we can avoid these getClassAttribs calls.
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:5349]: @jkotas do you think it makes sense to change this in `compareTypesForCast`? From what I understand, the current semantics of `compareTypesForCast` is "if I have an object of exact type A, can it be cast to type B" -- which is a meaningless question when `A` is an interface type. Alternatively, do you have any suggestions on better ways to check what we're after here?

Comment (jakobbotsch) [reply src/coreclr/jit/importer.cpp:5349]: Hmm, looks like we fold `Type.IsAssignableFrom`/`Type.IsAssignableTo` using `compareTypesForCast` as well, so seems like we cannot really change its behavior.

**Comment by jakobbotsch** [comment]
@MihuBot

File: src/coreclr/jit/importer.cpp (hunk 1 above)
Comment (jkotas) [reply src/coreclr/jit/importer.cpp:5349]: > do you think it makes sense to change this in compareTypesForCast?

Yes, I think this logic should be better done on the EE side.

> looks like we fold Type.IsAssignableFrom/Type.IsAssignableTo using compareTypesForCast as well, so seems like we cannot really change its behavior.

You can add a flag to the `compareTypesForCast` or add a new JIT/EE interface method.

**Pull request closed without merging**